californium-core/src/main/java/org/eclipse/californium/core/network/UdpMatcher.java

```diff
@@ -396,11 +398,12 @@ public void completed(final Exchange exchange) {
 
 		@Override
 		public void contextEstablished(final Exchange exchange) {
-
-			if (exchange.getRequest() != null) {
-				observationStore.setContext(exchange.getRequest().getToken(), exchange.getCorrelationContext());
+			Request currentRequest = exchange.getCurrentRequest(); 
+			Request request = exchange.getRequest(); 
+			if (request == currentRequest && request != null && request.isObserve()) {
+				observationStore.setContext(request.getToken(), exchange.getCorrelationContext());
```

Comment:
We check that `request == currentRequest`.
This happens only for blockwise transfer.
For block2 it's ok as this avoid to set context for each block2 request.
For block1 this means that correlation context will only be set  when the request will be completely sent. I think this is a problem. I think we want to set the correlation as soon as possible.

Should we just avoid to call `contextEstablished` if correlation context is already set or just when it changed  ?
I mean in Exchange :
```java
public void setCorrelationContext(final CorrelationContext ctx) {
    if (correlationContext != ctx) { // or correlationContext != null depending the behavior we want
       	correlationContext = ctx;
        ExchangeObserver obs = this.observer;
        if (obs != null) {
            obs.contextEstablished(this);
        }
    }
}
```